src/Cabal2JSON.hs

```diff
@@ -353,24 +358,56 @@ instance HasCodec Mixin where
         <*> requiredField' "include-renaming" .= mixinIncludeRenaming
 
 instance HasCodec LegacyExeDependency where
-  -- TODO: ?
-  codec = undefined
+  codec =
+    object "LegacyExeDependency" $
+      LegacyExeDependency
+        <$> requiredField' "package-name" .= (\(LegacyExeDependency s _) -> s)
+        <*> requiredField' "include-renaming" .= (\(LegacyExeDependency _ vr) -> vr)
 
 instance HasCodec ExeDependency where
-  -- TODO: ?
-  codec = undefined
+  codec =
+    object "ExeDependency" $
+      ExeDependency
+        <$> requiredField' "package-name" .= (\(ExeDependency a _ _) -> a)
+        <*> requiredField' "executable-component-name" .= (\(ExeDependency _ b _) -> b)
+        <*> requiredField' "version-range" .= (\(ExeDependency _ _ c) -> c)
 
 instance HasCodec PkgconfigDependency where
-  -- TODO: ?
-  codec = undefined
+  codec =
+    object "PkgconfigDependency" $
+      PkgconfigDependency
+        <$> requiredField' "pkgconfigName" .= (\(PkgconfigDependency name _) -> name)
+        <*> requiredField' "pkgconfigVersionRange" .= (\(PkgconfigDependency _ version) -> version)
+
+instance HasCodec PkgconfigName where
+  codec = dimapCodec mkPkgconfigName unPkgconfigName codec
+
+instance HasCodec PkgconfigVersionRange where
+  -- TODO add other versions
+  codec = stringConstCodec $ fromList [(PcAnyVersion , "any")]
 
 instance HasCodec Language where
-  -- TODO: ?
-  codec = undefined
+  codec = parseAlternative validLanguages unknownLanguage
+    where validLanguages = stringConstCodec $ fromList [
+              (Haskell98, "Haskell98")
+            , (Haskell2010, "Haskell2010")
+            ]
+          unknownLanguage = object "UnknownLanguage" $ UnknownLanguage <$> requiredField' "language" .= (\(UnknownLanguage k) -> k)
 
 instance HasCodec Extension where
-  -- TODO: ?
-  codec = undefined
+  codec = matchChoicesCodec[(\case
+      (EnableExtension ex) -> Just (EnableExtension ex)
+      _ -> Nothing
+    , object "EnableExtension" $ EnableExtension <$> requiredField' "name" .= (\(EnableExtension k) -> k)
+    ), (\case
+      (DisableExtension ex) -> Just (DisableExtension ex)
+      _ -> Nothing
+    , object "DisableExtension" $ DisableExtension <$> requiredField' "name" .= (\(DisableExtension k) -> k)
+    )]
+    (object "UnknownExtension" $ UnknownExtension <$> requiredField' "name" .= (\(UnknownExtension s) -> s))
```

Comment:
This seems bad. Is there a better way to do this?

Reply:
You use `eitherCodec` with tuples and then `dimapCodec`.

Reply:
Sry, I still don't get how this works. What is `eitherCodec` with tuples?

`Extension` has 3 constructors but I can only differentiate 2 with `either`.

```
instance HasCodec Extension where
  codec = dimapCodec f g $
       disjointEitherCodec
         (codec :: JSONCodec KnownExtension)
         (codec :: JSONCodec String)
     where
       f = \case
         Left e -> EnableExtension e
         Right e -> UnknownExtension e
       g = \case
         EnableExtension e -> Left e
         DisableExtension e -> ???
         UnknownExtension e -> Right e
```

Reply:
Then you nest two `eitherCodec`s.
Here's an example:
https://github.com/NorfairKing/intray/blob/b1b4c14c1a4e419d2a1bf96d757780981229a1e2/intray-api/src/Intray/API/Protected/Account/Types.hs#L49-L74

Reply:
Done